### ISSUE TITLE
If there is no<a> wrapper, the image cannot be opened

### ISSUE THREAD

**Issue by otary** [dimsemenov/PhotoSwipe#2051]
```
<div id="test3">
            <img class="img2"
                 src="https://img2.baidu.com/it/u=3329909248,3799019568&fm=253&fmt=auto&app=138&f=JPEG?w=889&h=500"
            />
            <img class="img2"
                 src="https://img2.baidu.com/it/u=3329909248,3799019568&fm=253&fmt=auto&app=138&f=JPEG?w=889&h=500"
            />
            <img class="img2"
                 src="https://img2.baidu.com/it/u=3329909248,3799019568&fm=253&fmt=auto&app=138&f=JPEG?w=889&h=500"
            />
 </div>
```

```
const photoSwipeLightbox = new PhotoSwipeLightbox({
                 gallery: '#test3',
                children: '.img2',
                pswpModule: () => PhotoSwipe
});
photoSwipeLightbox.init();
```

As shown above, clicking on the image can open it, but it cannot display the image！

**Comment by arnowelzel** [comment]
Without an <a> wrapper there is nothing to click on and users won't expect that clicking an element which is **not** a link to something or a button element will do something. Especially since the mouse pointer won't indicate this as well. You may need to build a collection of all images first yourself and add a click handler for each image which will then call the respective method in PhotoSwipe to open a lightbox for the specific image.

**Comment by KnIfER** [comment]
you can bypass the `"A"===t.tagName` check inside of `photoswipe.umd.min.js ` and `photoswipe-lightbox.umd.min.js`

**Comment by arnowelzel** [comment]
And by the way - it is not the visible images which are display in the lightbox it is the the **links**. The visible images should just be smaller thumbnails and not the final images.

**Comment by Anton-Evstigneev** [comment]
Such tight coupling with anchor tag results in bad UX when user tries to click on image while page's still loading. So instead of opening zoom overlay, user is navigated away from the site to an empty page with just one image.

**Comment by arnowelzel** [comment]
About the "bad UX" comment: when the user clicks on an image before the loading process has finished, the script are not loaded yet. That's the reason why the browser will show the image instead of opening the lightbox. But I would like to see your alternative how it should be handled instead. Any live example? Also keep in mind, that event handlers need to be added to the images for PhotoSwipe. I'd like to learn how to do this while the document is still loading and the DOM is not built completely yet.

And in general - maybe there is misunderstanding how image galleries on websites work. Except a very few execptions, the images which are **visible** in the website are **not** the ones which are displayed in the lightbox.

So the usual pattern is:

```
<a href="full-resolution-image.jpg"><img src="preview-image.jpg" alt="alternative text" /></a>
```

Also keep in mind the **semantic** here: the anchor tells the browser "here is something the user can open". This is important! By adding an anchor, the browser will also display a different pointer when hovering with the mouse and makes the element focusable. When the user is not able to use a mouse but navigates by switchting from one focusable element to the next, this is also important: an image itself is not focusable - only anchors with a reference and input elements are.

Also keep in mind: how to distinguish between images which are just decorative and others which are intended to be clicked by the user? This is also indicated by adding an anchor element - as an indication "this image links to something you can open".

In my opinion I would close this issue as "won't fix" since the way how PhotoSwipe works is the logical and expected behaviour: only those images will be used for the lightbox which would be opened anyway when them, even without JavaScript and PhotoSwipe. If you want to have a custom solution, you still can create your own frontend script which adds all images to a collection and pass that to PhotoSwipe and add event handlers to the collected images. You can also generate the script code in the backend and pass that along with "onclick" handlers hardcoded in the image tags.

**Comment by tennox** [comment]
I also would prefer something like this:
```vue
<div class="gallery-image cursor-pointer" role="button">
  <img ... >
</div>
```
([`role=button docs`](https://www.w3.org/WAI/PF/aria/roles#button))
but I get your point with using a semantic tag like `<a>`, so here is a way to address the
> bad UX when user tries to click on image while page's still loading
```
<a href="javascript:void(0);" data-pwsp-src="imageUrl">
```
this makes the link just do nothing when gallery code is not loaded yet

**Comment by arnowelzel** [comment]
> ...but I get your point with using a semantic tag like `<a>`, so here is a way to address the
> 
> > bad UX when user tries to click on image while page's still loading
> 
> ```
> <a href="javascript:void(0);" data-pwsp-src="imageUrl">
> ```
> 
> this makes the link just do nothing when gallery code is not loaded yet

Yes, this would solve the issue tha timages can not be clicked before the DOM is completely loaded.

But I am not sure if this is "good UX" since it creates a new problem: If JavaScript is disabled - maybe because the user does not want this for security reasons or there are security policies which forbid JavaScript by default - image links don't work at all.

What you describe as "bad UX" is just the fact, that images may not be opened in the PhotoSwipe lightbox if a user clicks too early. However images will be opened in any case, even if there is no JavaScript at all - either in a lightbox or as a single image. This is still better than keeping everybody from opening an image if there is no JavaScript available. Yes, having JavaScript avialable is expected nowadays, but good web design should always keep graceful degradation in mind which is a very important thing for accessibility too.

I also doubt, that handling the consequences of slow loading websites is something which should be addressed in PhotoSwipe. If a website takes significantly longer than a few seconds to load under normal conditions (which means at least 10 MBit/s downstream which should be the baseline for most users), then there a more serious issues that should be fixed in the server side.

**Comment by Andrew-web-coder** [comment]
Interestingly, Fancybox works fine without a wrapping element - https://jsfiddle.net/hdp70a6m/

**Comment by arnowelzel** [comment]
Yes, you can capture clicks on **every** element! But this does not mean, that this is a good idea.

IMG elements without a A wrapper can **not** be focused and therefore this is **not** barrier free!

**Comment by tennox** [comment]
> IMG elements without an A wrapper around it can not be focused and therefore this is not barrier free!

How about setting [`tabindex`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Keyboard) ?

**Comment by arnowelzel** [comment]
`tabindex` would work indeed. But then you have exactly the same behaviour as before - and this makes no sense. And `tabindex` is also not ideal if the images are not the only focusable element - because then you should give focusable elements on your site the same tabindex to make sure that the tab order is not changed.

My opinion about this is still to just keep the wrapper `<a>` where it is. It is accessible and works. I already explained everything in detail [earlier](https://github.com/dimsemenov/PhotoSwipe/issues/2051#issuecomment-2033899521).

**Comment by wisniewski94** [comment]
TLDR; This element should be a button, which by its function is. 

@arnowelzel 
> It is accessible and works. I already explained everything in detail 
> Also keep in mind the semantic here: the anchor tells the browser "here is something the user can open".

It's not accessible. Please check the attached video.
1. The link is announced with a terrible name, implying an element leads to an unknown page.
2. The clicked link will open dialogue which is even more confusing. As visually impaired user (which I am not) I expected to hear a new page title.
3. Sometimes screen reader will announce the element was visited (how can you visit a popup?).

https://github.com/user-attachments/assets/f6228d92-555d-45d2-aa27-3d8e4cef32ac

This is quite a serious issue in terms of upcoming EU directive and similar regulations across the globe.

On the other hand, it will impact crawlers affecting SEO. I suspect these links could be caught by Google's tools such as Search Console and be marked as errors. I believe there could be more reasons to use HTML elements according to their original purpose that I can't think of right now. To summarize it - there are strong grounds to call this design an anti-pattern which has nothing to do with good UX practices or accessibility.

**Comment by arnowelzel** [comment]
You can always create your own frontend code to build whatever kind of lightbox collection you want. If you believe, that  using <button> elements works better to open thumbnails - freel free to build it this way. At the moment I don't even know, how to build a <button> element which displays a thumbnail image and can be clicked as well **and** still links to the image in case JavaScript is disabled.

**Comment by arnowelzel** [comment]
> On the other hand, it will impact crawlers affecting SEO. I suspect these links could be caught by Google's tools such as Search Console and be marked as errors. I believe there could be more reasons to use HTML elements according to their original purpose that I can't think of right now. To summarize it - there are strong grounds to call this design an anti-pattern which has nothing to do with good UX practices or accessibility.

Please show your solution to this. How would you create a picture gallery where one can click images which and where you can open the images by clicking the thumbnails without JavaScript as fallback.

Also I don't see crawler errors in websites using PhotoSwipe (the biggest site I am hosting has around 6000 images in galleries where the thumbnails are linked to the bigger versions using "a" elements and so far at least Google did not complain about that).

**Comment by arnowelzel** [comment]
There is no "lightbox element" - there is just a link to a bigger image and PhotoSwipe intercepts the action "open the link to the image" so it will be displayed in a lightbox.

Again: present your code! Show how you would to it! How should a image gallery look like which should be handled by PhotoSwipe as well?

**Comment by arnowelzel** [comment]
Example: https://codeberg.org/askaaron/lightbox-photoswipe/src/tag/5.4.1/src/js/frontend.js (moved from https://github.com/arnowelzel/lightbox-photoswipe/blob/5.4.1/src/js/frontend.js)

This is a custom script using PhotoSwipe 5 to display images using PhotoSwipe in WordPress **if** they are indicated as "image should be displayed in a lightbox" by the backend.

Real life use examples:

https://arnowelzel.de/en/magura-rim-brakes-old-and-new
https://arnowelzel.de/en/minerve-2016

I would really like to learn how the image galleries or individual images should look like using "button" element instead of "a".

And keep in mind: not the thumbnail is opened in the lightbox but ANOTHER image!

**Comment by arnowelzel** [comment]
Or to make it a bit easier - what should be the alternative to this code:

```
<a href="big-image-1.jpg"><img src="image-thumbnail-1.jpg"></a>
<a href="big-image-2.jpg"><img src="image-thumbnail-2.jpg"></a>
<a href="big-image-3.jpg"><img src="image-thumbnail-3.jpg"></a>
```

Please show an example, how this could be done better, barrier free and still working without JavaScript enabled, so the big images still can opened by clicking the thumbnails.

**Comment by juji** [comment]
https://pele.jujiplay.com/

```html
<div class="inside-module__iMU7oW__container">
  <h2>Inside Pelé</h2>
  <div class="inside-module__iMU7oW__images">
    <img 
      loading="lazy" 
      data-zoomable="/images/inside/63dd9c47c8261_s1.webp" 
      data-zoomable-width="600" 
      data-zoomable-height="388" 
      data-zoomable-group="inside" 
      src="/images/inside/63dd9c47c8261_s1.webp" 
      alt="Inside Pelé Book Page 1" 
      width="600" 
      height="388"
      decoding="async" />
    <img 
      loading="lazy" 
      data-zoomable="/images/inside/63dd9c6c807b4_s2.webp" 
      data-zoomable-width="600" 
      data-zoomable-height="386" 
      data-zoomable-group="inside" 
      src="/images/inside/63dd9c6c807b4_s2.webp" 
      alt="Inside Pelé Book Page 2"
      width="600" 
      height="386" 
      decoding="async"
    />
    <!-- ... more images ... -->
  </div>
</div>
```

```js
const groups: Record<string, { parent: Element | null }> = {};

document.querySelectorAll('[data-zoomable]').forEach((element) => {
  
  const group =
    element.getAttribute('data-zoomable-group') ||
    Math.random().toString(36).substring(2, 15);

  if (!groups[group]) {
    groups[group] = {
      parent: null,
    };
  }

  if (!groups[group].parent) {
    groups[group].parent = element.parentElement;
  }
});

for (const groupId in groups) {
  const group = groups[groupId];
  if (!group.parent) continue;

  const lightbox = new PhotoSwipeLightbox({
    gallery: group.parent as HTMLElement,
    children: '[data-zoomable]',
    pswpModule: () => import('@juji/photoswipe'), // import('photoswipe') will be the same
  });

  lightbox.addFilter('domItemData', (itemData, el) => {
    itemData.src = el.getAttribute('data-zoomable') || '';
    itemData.w = parseInt(
      el.getAttribute('data-zoomable-width') || '0',
      10
    );
    itemData.h = parseInt(
      el.getAttribute('data-zoomable-height') || '0',
      10
    );
    itemData.msrc = el.getAttribute('src') || '';
    itemData.thumbCropped = false;
    return itemData;
  });

  lightbox.init();
  // lightboxRef.current.push(lightbox);
}
```

https://github.com/user-attachments/assets/0f3aca4a-252e-450d-a222-d2eeaa475f60

**Comment by arnowelzel** [comment]
Without JavaScript the images can not be opened.

**Comment by juji** [comment]
@otary this is a way to open up without the a tag...

however, yes @arnowelzel i think we should make it open without javascript, to include non-js environment

**Comment by arnowelzel** [comment]
Yes - this was my point from the beginning:

`<a href="...">` indicates "there is a link to something". Elements which are not inside `<a>` are not intended to be opened, even when it is possible to "emulate" this by adding a JavaScript event handler for mouse clicks.

And even if you add an event handler - you have to "emulate" all possible actions: mouse click, focusing using the [Tab] key, opening using [Space] or [Enter] (depending on the browser) and touch events. The solution above does not take care of any accessibility related issues - like keyboard navigation for example.

The second problem: images in the page should be thumbnails and link to the full sized images. Just having a gallery with full sized images which can be enlarged using a lightbox makes nut much sense. So for example `/images/inside/63dd9c47c8261_s1.webp` is the "zoomable" image as well as the real image inside the page.

So the idea is to have smaller thumbnails with larger images as link:

```
<a href="BIGIMAGE"><img src="THUMBNAIL" /></a>
```

First of all this is accessible - people can use their keyboard to navigate as well (also see https://wordpress-demo.arnowelzel.de/lightbox-with-photoswipe-5/ as example).

Then you load only smaller images as thumbnails and only transfer the bigger images when people want to see them.

And finally - even if JavaScript is not possible, the bigger images can still be opened, just without a lightbox.

**Comment by xav-ie** [comment]
Hey, I know this is old thread, but what do people think of this?

SSR, initial load:
```
<div aria-label="A pretty flower" class="stack">
  <img alt="A pretty flower" aria src="small-img.jpeg"/>
  <a href="Open full size image"></a>
  <button class="display-none" onClick={openFullSize}><span class="sr-only">Open full size image</span>
<div>
```

To display this, you would use CSS-grid stacking method: https://joyofcode.xyz/stacking-items-with-css-grid

This way, the link stretches over the entire visible image.

Then, on hydration/js, you simply:
1. add `display-none` class to the link
2. remove `display-none` class from the button


I think this has a few advantages. I am noticing in general it is pretty much always a bad idea to try and wrap things inside of things when trying to do anything good for a11y.

Here is a link to a codepen where I try to explore it more:
https://codepen.io/xav-ie/pen/YPWZwzb

I explain a bit more detailed and specific advantages over link approach there. I encourage you to use a screen reader and let me know what you think.
